py-polars/polars/selectors.py

```diff
@@ -31,98 +32,129 @@
         from typing_extensions import Self
 
 
-def _expand_selectors(
-    frame: DataFrame | LazyFrame, items: Any, *more_items: Any
-) -> list[Any]:
+def is_selector(obj: Any) -> bool:
     """
-    Expand any selectors to column names in the given input.
+    Indicate whether the given object/expression is a selector.
 
-    Non-selector values are left as-is.
+    Examples
+    --------
+    >>> from polars.selectors import is_selector
+    >>> import polars.selectors as cs
+    >>> is_selector(pl.col("colx"))
+    False
+    >>> is_selector(cs.first() | cs.last())
+    True
+    """
+    # note: don't want to expose the "_selector_proxy_" object
+    return isinstance(obj, _selector_proxy_)
+
+
+def expand_selector(
+    target: DataFrame | LazyFrame | Mapping[str, PolarsDataType], selector: SelectorType
+) -> tuple[str, ...]:
+    """
+    Expand a selector to column names with respect to a specific frame or schema target.
+
+    Parameters
+    ----------
+    target
+        A polars DataFrame, LazyFrame or schema.
+    selector
+        An arbitrary polars selector (or compound selector).
 
     Examples
     --------
-    >>> from polars.selectors import _expand_selectors
+    >>> from polars.selectors import expand_selector
     >>> import polars.selectors as cs
     >>> df = pl.DataFrame(
     ...     {
-    ...         "colw": ["a", "b"],
     ...         "colx": ["x", "y"],
     ...         "coly": [123, 456],
     ...         "colz": [2.0, 5.5],
     ...     }
     ... )
-    >>> _expand_selectors(df, ["colx", cs.numeric()])
-    ['colx', 'coly', 'colz']
-    >>> _expand_selectors(df, cs.string(), cs.float())
-    ['colw', 'colx', 'colz']
-
+    >>> expand_selector(df, cs.numeric())
+    ('coly', 'colz')
+    >>> expand_selector(df, cs.first() | cs.last())
+    ('colx', 'colz')
+    >>> expand_selector(df, ~(cs.first() | cs.last()))
+    ('coly',)
     """
-    expanded: list[Any] = []
-    for item in (
-        *(
-            items
-            if isinstance(items, Collection) and not isinstance(items, str)
-            else [items]
-        ),
-        *more_items,
-    ):
-        if is_selector(item):
-            selector_cols = selector_column_names(frame, item)
-            expanded.extend(selector_cols)
-        else:
-            expanded.append(item)
-    return expanded
-
+    if isinstance(target, Mapping):
+        from polars.dataframe import DataFrame
 
-def is_selector(obj: Any) -> bool:
-    """
-    Indicate whether the given object/expression is a selector.
+        target = DataFrame(schema=target)
+    else:
+        target = target.clear()
```

Comment:
Why does it have to be cleared?

Reply:
Ah! It doesn't; that's a leftover from the previous iteration that I meant to ditch now I'm updating it - fixed 👍